lib/demuxer.ex

```diff
@@ -0,0 +1,161 @@
+defmodule Membrane.Ogg.Demuxer do
+  use Membrane.Filter
+  alias Membrane.{Buffer, RemoteStream, Opus}
+
+  def_input_pad :input,
+    availability: :always,
+    mode: :pull,
+    demand_unit: :buffers,
+    accepted_format: any_of(RemoteStream)
+
+  def_output_pad :output,
+    availability: :on_request,
+    mode: :pull,
+    accepted_format: %RemoteStream{type: :packetized, content_format: Opus}
+
+  defmodule State do
+    @type t :: %__MODULE__{
+            actions_buffer: list,
+            parser_acc: binary,
+            phase: :all_outputs_linked | :awaiting_linking,
+            continued_packets: %{}
+          }
+
+    defstruct actions_buffer: [],
+              parser_acc: <<>>,
+              phase: :awaiting_linking,
+              continued_packets: %{}
+  end
+
+  @impl true
+  def handle_init(_ctx, _options) do
+    {[], %State{}}
+  end
+
+  @impl true
+  def handle_playing(_context, state) do
+    {[{:demand, :input}], state}
+  end
+
+  @impl true
+  def handle_pad_added(Pad.ref(:output, id), _context, state) do
+    {[
+       stream_format:
+         {Pad.ref(:output, id), %RemoteStream{type: :packetized, content_format: Opus}}
+     ], %State{state | phase: :all_outputs_linked}}
+  end
+
+  @impl true
+  def handle_process(:input, %Buffer{payload: bytes}, context, state) do
+    unparsed = state.parser_acc <> bytes
+
+    {parsed, unparsed, continued_packets} =
+      Membrane.Ogg.Parser.parse(unparsed, state.continued_packets)
+
+    state = %State{state | parser_acc: unparsed}
+    state = %State{state | continued_packets: continued_packets}
+
+    {actions, state} = process_packets({[], context, state}, parsed)
+
+    demand_if_not_blocked({actions, state})
+  end
+
+  @impl true
+  def handle_demand(Pad.ref(:output, _id), _size, :buffers, context, state)
+      when state.phase == :all_outputs_linked do
+    {[], state}
+    |> reclassify_buffer_actions(context)
+    |> demand_if_not_blocked()
+  end
+
+  @impl true
+  def handle_demand(Pad.ref(:output, _id), _size, :buffers, _context, state) do
+    {:ok, state}
+  end
+
+  defp process_packets({actions, context, state}, packets_list) do
+    {actions, _context, state} =
+      Enum.reduce(packets_list, {actions, context, state}, &process_packet/2)
+
+    {actions, state}
```

Comment:
You don't modify `context` so it can be excluded from acc

```suggestion
    Enum.reduce(packets, {actions, state}, &process_packet/2)
```    

After this small refactor we can inline `process_packets` function as it is used only in one place

Reply:
The context is modified here https://github.com/membraneframework-labs/membrane_ogg_plugin/pull/1/files#diff-30f1ab064b7d41422f5a7751b2e2d5eb36b10d3dceecbeff187424909dcceef6R134 in `classify_buffer_action` function

Reply:
Yeah but it has no effect as everything in Elixir is immutable. Context is also internal Membrane structure that  can only be read - it's impossible to return context from any membrane callback to modify it